### PR TITLE
Added conical and confined axis constraints, relevent Json files, and…

### PR DESCRIPTION
**Pull request by reidchristopher**
… tests for the new constraints.

### CODE REVIEW

File: trajopt/include/trajopt/file_write_callback.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#include <trajopt_utils/eigen_conversions.hpp>
+#include <tesseract_core/basic_kin.h>
+#include <tesseract_core/basic_env.h>
+#include <trajopt/plot_callback.hpp>
+#include <functional>
+#include <set>
+#include <fstream>
+#include <string>
+#include <stdio.h>
+#include <Eigen/Core>
```
Comment (Jmeyer1292) [inline trajopt/include/trajopt/file_write_callback.hpp]: Don't put includes in a public header that aren't needed _in the header_. Move as much to the source file as possible.

File: trajopt/include/trajopt/file_write_callback.hpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+typedef std::shared_ptr<std::ofstream> ofstreamPtr;
+
+using namespace Eigen;
+using namespace util;
+using namespace std;
```
Comment (Jmeyer1292) [inline trajopt/include/trajopt/file_write_callback.hpp]: Is this meant to be a publicly accessed header? If so, do not put using directives in it. You're effectively also making the decision to use the same namespaces on behalf of all of your clients. Put it in the source file.

File: trajopt/include/trajopt/file_write_callback.hpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+namespace trajopt
+{
+
+  Optimizer::Callback WriteCallback(string file_name,TrajOptProbPtr prob, vector<string> joint_names,
```
Comment (Jmeyer1292) [inline trajopt/include/trajopt/file_write_callback.hpp]: Is the lack of references in this function signature intentional? Why not pass joint names as a `const std::vector<std::string>&`?

Comment (Jmeyer1292) [reply trajopt/include/trajopt/file_write_callback.hpp]: I'm sometimes lazy about this (and I'm not sure it matters for 64bit machines), but Eigen types require certain memory alignment for their SIMD acceleration to function correctly.

So when writing `std::vector<Eigen::Affine3d>`, you should probably write instead: `std::vector<Eigen::Affine3d, Eigen::aligned_allocator<Eigen::Affine3d>>` instead. There's a package in ROS called  `eigen_stl_containers` that provides some aliases for you. It's probably better to define your own using templated aliases.

File: trajopt/include/trajopt/file_write_callback.hpp (hunk 2 above)
Comment (reidchristopher) [reply trajopt/include/trajopt/file_write_callback.hpp]: Definitely needs to be changed. Didn't understand the extent of what this did back then

File: trajopt/include/trajopt/file_write_callback.hpp (hunk 3 above)
Comment (reidchristopher) [reply trajopt/include/trajopt/file_write_callback.hpp]: I cannot remember if the pass by value was purposeful. Can you explain to me what passing Eigen's allocator does and why it is preferable?

**Pull request closed without merging**